In [72]:
import numpy as np 
import pandas as pd 
import seaborn as sns 
import matplotlib.pyplot as plt  


In [73]:
df = pd.read_csv("Bengaluru_House_Data.csv")
df

,area_type,availability,location,size,society,total_sqft,bath,balcony,price
0,Super built-up Area,19-Dec,Electronic City Phase II,2 BHK,Coomee,1056,2.0,1.0,39.07
1,Plot Area,Ready To Move,Chikka Tirupathi,4 Bedroom,Theanmp,2600,5.0,3.0,120.00
2,Built-up Area,Ready To Move,Uttarahalli,3 BHK,NaN,1440,2.0,3.0,62.00
3,Super built-up Area,Ready To Move,Lingadheeranahalli,3 BHK,Soiewre,1521,3.0,1.0,95.00
4,Super built-up Area,Ready To Move,Kothanur,2 BHK,NaN,1200,2.0,1.0,51.00
...,...,...,...,...,...,...,...,...,...
13315,Built-up Area,Ready To Move,Whitefield,5 Bedroom,ArsiaEx,3453,4.0,0.0,231.00
13316,Super built-up Area,Ready To Move,Richards Town,4 BHK,NaN,3600,5.0,NaN,400.00
13317,Built-up Area,Ready To Move,Raja Rajeshwari Nagar,2 BHK,Mahla T,1141,2.0,1.0,60.00
13318,Super built-up Area,18-Jun,Padmanabhanagar,4 BHK,SollyCl,4689,4.0,1.0,488.00


In [74]:
df.shape

(13320, 9)

In [75]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  str    
 1   availability  13320 non-null  str    
 2   location      13319 non-null  str    
 3   size          13304 non-null  str    
 4   society       7818 non-null   str    
 5   total_sqft    13320 non-null  str    
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), str(6)
memory usage: 1.6 MB


In [76]:
df.duplicated().sum()

529

In [77]:
df.isna().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

# data overview 

In [78]:
df["area_type"].value_counts()

area_type
Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2025
Carpet  Area              87
Name: count, dtype: int64

In [79]:
df["availability"].value_counts().sum

<bound method Series.sum of availability
Ready To Move    10581
18-Dec             307
18-May             295
18-Apr             271
18-Aug             200
                 ...  
15-Aug               1
17-Jan               1
16-Nov               1
16-Jan               1
14-Jul               1
Name: count, Length: 81, dtype: int64>

In [80]:
df["location"].value_counts()

location
Whitefield                                         540
Sarjapur  Road                                     399
Electronic City                                    302
Kanakpura Road                                     273
Thanisandra                                        234
                                                  ... 
Pattegarhpalya                                       1
Tilak Nagar                                          1
12th cross srinivas nagar banshankari 3rd stage      1
Havanur extension                                    1
Abshot Layout                                        1
Name: count, Length: 1305, dtype: int64

In [81]:
df["location"].nunique()

1305

we have 1 row with null values in the location column and there are 1305 unique values so  we can drop the row 

In [82]:
df["size"].value_counts() 
# we can remove BHK and bedroom from the  columns to convert them to nummeric   


size
2 BHK         5199
3 BHK         4310
4 Bedroom      826
4 BHK          591
3 Bedroom      547
1 BHK          538
2 Bedroom      329
5 Bedroom      297
6 Bedroom      191
1 Bedroom      105
8 Bedroom       84
7 Bedroom       83
5 BHK           59
9 Bedroom       46
6 BHK           30
7 BHK           17
1 RK            13
10 Bedroom      12
9 BHK            8
8 BHK            5
11 BHK           2
11 Bedroom       2
10 BHK           2
27 BHK           1
19 BHK           1
16 BHK           1
43 Bedroom       1
14 BHK           1
12 Bedroom       1
13 BHK           1
18 Bedroom       1
Name: count, dtype: int64

In [83]:
df["society"].value_counts()

society
GrrvaGr    80
PrarePa    76
Sryalan    59
Prtates    59
GMown E    56
           ..
Maa 5a      1
AeisePV     1
SJovest     1
ThhtsV      1
RSntsAp     1
Name: count, Length: 2688, dtype: int64

In [84]:
df["society"].nunique()

2688

In [85]:
df.describe()

,bath,balcony,price
count,13247.000000,12711.000000,13320.000000
mean,2.692610,1.584376,112.565627
std,1.341458,0.817263,148.971674
min,1.000000,0.000000,8.000000
25%,2.000000,1.000000,50.000000
50%,2.000000,2.000000,72.000000
75%,3.000000,2.000000,120.000000
max,40.000000,3.000000,3600.000000


In [86]:
df.describe(include="object")

C:\Users\pushk\AppData\Local\Temp\ipykernel_30572\702825166.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,area_type,availability,location,size,society,total_sqft
count,13320,13320,13319,13304,7818,13320
unique,4,81,1305,31,2688,2117
top,Super built-up Area,Ready To Move,Whitefield,2 BHK,GrrvaGr,1200
freq,8790,10581,540,5199,80,843


here total_sqft is in str which we need to convert it to the  integers 

# Data cleaning 

In [87]:
df["total_sqft"].value_counts()

total_sqft
1200           843
1100           221
1500           205
2400           196
600            180
              ... 
1020 - 1130      1
2758             1
1133 - 1384      1
774              1
4689             1
Name: count, Length: 2117, dtype: int64

In [88]:
def clean_total_sqft(x):
    if pd.isna(x):
        return np.nan

    x = str(x).strip()
    if "perch" in x.lower():
        num_str = x.lower().replace("perch", "").strip()
        return pd.to_numeric(num_str) * 272.25
    if "sq. yards" in x.lower():
        num_str = x.lower().replace("sq. yards", "").strip()
        return pd.to_numeric(num_str) * 9.0
    if "acre" in x.lower():
        num_str = x.lower().replace("acres", "").replace("acre", "").strip()
        return pd.to_numeric(num_str) * 43560.0
    if "sq. meter" in x.lower():
        num_str = x.lower().replace("sq. meter", "").strip()
        return pd.to_numeric(num_str) * 10.7639
    if "cent" in x.lower():
        num_str = x.lower().replace("cents", "").replace("cent", "").strip()
        return pd.to_numeric(num_str) * 435.6
    if "ground" in x.lower():
        num_str = x.lower().replace("grounds", "").replace("ground", "").strip()
        return pd.to_numeric(num_str) * 2400.0
    if "guntha" in x.lower():
        num_str = x.lower().replace("guntha", "").strip()
        return pd.to_numeric(num_str) * 1089.0
    if "-" in x:
        parts = x.split("-")
        v1 = pd.to_numeric(parts[0].strip())
        v2 = pd.to_numeric(parts[1].strip())
        return (v1 + v2) / 2.0
    return pd.to_numeric(x)

In [89]:
df["total_sqft"] = df["total_sqft"].apply(clean_total_sqft)

In [90]:
df["total_sqft"].value_counts()

total_sqft
1200.0    843
1100.0    221
1500.0    206
2400.0    197
600.0     180
         ... 
2395.0      1
2758.0      1
1258.5      1
774.0       1
4689.0      1
Name: count, Length: 2034, dtype: int64

In [91]:
df.loc[df["location"].isna()]

,area_type,availability,location,size,society,total_sqft,bath,balcony,price
568,Super built-up Area,Ready To Move,NaN,3 BHK,Grare S,1600.0,3.0,2.0,86.0


In [92]:
df.loc[df["society"]=="Grare S"]

,area_type,availability,location,size,society,total_sqft,bath,balcony,price
568,Super built-up Area,Ready To Move,NaN,3 BHK,Grare S,1600.0,3.0,2.0,86.0
12238,Carpet Area,Ready To Move,Anantapura,3 BHK,Grare S,1600.0,3.0,2.0,77.0


In [93]:
df["location"] = df["location"].fillna("Anantapura")

In [94]:
df.loc[df["location"].isna()]

,area_type,availability,location,size,society,total_sqft,bath,balcony,price


In [103]:
df["size"].isna().sum()

16

In [ ]:
df.loc[df["size"].isna()]

,area_type,availability,location,size,society,total_sqft,bath,balcony,price
579,Plot Area,Immediate Possession,Sarjapur Road,NaN,Asiss B,1800.0,NaN,NaN,34.185
1775,Plot Area,Immediate Possession,IVC Road,NaN,Orana N,3817.0,NaN,NaN,124.000
2264,Plot Area,Immediate Possession,Banashankari,NaN,NaN,2400.0,NaN,NaN,460.000
2809,Plot Area,Immediate Possession,Sarjapur Road,NaN,AsdiaAr,1800.0,NaN,NaN,28.785
2862,Plot Area,Immediate Possession,Devanahalli,NaN,Ajleyor,1950.0,NaN,NaN,46.800
5333,Plot Area,Immediate Possession,Devanahalli,NaN,Emngs S,3752.5,NaN,NaN,177.115
6423,Plot Area,Immediate Possession,Whitefield,NaN,SRniaGa,2324.0,NaN,NaN,26.730
6636,Plot Area,Immediate Possession,Jigani,NaN,S2enste,1500.0,NaN,NaN,25.490
6719,Plot Area,Immediate Possession,Hoskote,NaN,SJowsn,1730.0,NaN,NaN,28.545
7680,Plot Area,Immediate Possession,Kasavanhalli,NaN,NaN,5000.0,NaN,NaN,400.000


In [115]:
df = df[df["size"].notna()]

In [116]:
df["size"].isna().sum()

0

In [117]:
def extract_bhk_components(x):
    cols = ["bedrooms", "halls", "kitchens"]
    if pd.isna(x):
        return pd.Series([np.nan] * 3, index=cols)
    parts = str(x).strip().split()
    if not parts:
        return pd.Series([np.nan] * 3, index=cols)
    num = pd.to_numeric(parts[0], errors="coerce")
    unit = " ".join(parts[1:]).lower()
    if "rk" in unit:
        values = [0, 0, 1]
    elif any(k in unit for k in ["bhk", "bedroom"]):
        values = [num, 1, 1]
    else:
        values = [num, np.nan, np.nan]
    return pd.Series(values, index=cols)

In [118]:
df[["bedrooms", "halls", "kitchens"]] = df["size"].apply(extract_bhk_components)

In [123]:
df

,area_type,availability,location,size,society,total_sqft,bath,balcony,price,bedrooms,halls,kitchens
0,Super built-up Area,19-Dec,Electronic City Phase II,2 BHK,Coomee,1056.0,2.0,1.0,39.07,2,1,1
1,Plot Area,Ready To Move,Chikka Tirupathi,4 Bedroom,Theanmp,2600.0,5.0,3.0,120.00,4,1,1
2,Built-up Area,Ready To Move,Uttarahalli,3 BHK,NaN,1440.0,2.0,3.0,62.00,3,1,1
3,Super built-up Area,Ready To Move,Lingadheeranahalli,3 BHK,Soiewre,1521.0,3.0,1.0,95.00,3,1,1
4,Super built-up Area,Ready To Move,Kothanur,2 BHK,NaN,1200.0,2.0,1.0,51.00,2,1,1
...,...,...,...,...,...,...,...,...,...,...,...,...
13315,Built-up Area,Ready To Move,Whitefield,5 Bedroom,ArsiaEx,3453.0,4.0,0.0,231.00,5,1,1
13316,Super built-up Area,Ready To Move,Richards Town,4 BHK,NaN,3600.0,5.0,NaN,400.00,4,1,1
13317,Built-up Area,Ready To Move,Raja Rajeshwari Nagar,2 BHK,Mahla T,1141.0,2.0,1.0,60.00,2,1,1
13318,Super built-up Area,18-Jun,Padmanabhanagar,4 BHK,SollyCl,4689.0,4.0,1.0,488.00,4,1,1


In [127]:
df.loc[df["bath"].isna()]

,area_type,availability,location,size,society,total_sqft,bath,balcony,price,bedrooms,halls,kitchens
56,Built-up Area,20-Feb,Devanahalli,4 Bedroom,BrereAt,3210.0,NaN,NaN,192.000,4,1,1
81,Built-up Area,18-Oct,Hennur Road,4 Bedroom,Gollela,3203.5,NaN,NaN,224.500,4,1,1
224,Super built-up Area,19-Dec,Devanahalli,3 BHK,Jurdsig,1630.0,NaN,NaN,74.820,3,1,1
344,Super built-up Area,21-Dec,Kanakpura Road,1 BHK,PrarePa,525.0,NaN,NaN,21.530,1,1,1
669,Super built-up Area,18-Dec,JP Nagar,5 BHK,Pehtsa,5520.0,NaN,NaN,375.000,5,1,1
702,Super built-up Area,18-Dec,JP Nagar,5 BHK,Pehtsa,5600.0,NaN,NaN,548.500,5,1,1
801,Super built-up Area,18-Dec,JP Nagar,4 BHK,Pehtsa,4624.5,NaN,NaN,453.000,4,1,1
941,Super built-up Area,Ready To Move,Whitefield,4 Bedroom,PrOakSi,4348.5,NaN,NaN,304.000,4,1,1
1264,Built-up Area,18-May,Hennur,3 Bedroom,Asoilul,2264.0,NaN,NaN,155.000,3,1,1
1267,Super built-up Area,18-Jun,Yelahanka,3 BHK,Shalkri,1662.0,NaN,NaN,67.980,3,1,1


In [135]:
df["bath"] = df["bath"].fillna(-1)

In [134]:
df["bath"].isna().sum()

0

In [136]:
df["balcony"] = df["balcony"].fillna(-1)

In [137]:
df.isna().sum()

area_type          0
availability       0
location           0
size               0
society         5499
total_sqft         0
bath               0
balcony            0
price              0
bedrooms           0
halls              0
kitchens           0
dtype: int64

In [139]:
df=df.drop("size",axis=1)

In [141]:
df["society"].isna().sum()/len(df)

0.413334335538184

In [147]:
loc_to_society_mode = df.groupby("location")["society"].agg(
    lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan
)

In [148]:
df["society"] = df["society"].fillna(df["location"].map(loc_to_society_mode))

In [149]:
df["society"].isna().sum()

1413

In [150]:
df["society"] = df["society"].fillna("unknown")

In [151]:
df["society"].isna().sum()

0

In [155]:
df["area_type"].value_counts()

area_type
Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2009
Carpet  Area              87
Name: count, dtype: int64

# Univariate Analysis

In [153]:
num_col = df.describe().columns
cat_cols = df.describe(include="object").columns

C:\Users\pushk\AppData\Local\Temp\ipykernel_30572\2076273583.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = df.describe(include="object").columns


In [154]:
num_col,cat_cols

(Index(['total_sqft', 'bath', 'balcony', 'price', 'bedrooms', 'halls',
        'kitchens'],
       dtype='str'),
 Index(['area_type', 'availability', 'location', 'society'], dtype='str'))

In [156]:
for i in num_col:
    plt.figure(figsize=(5.6))
    sns.histplot(df[i])

TypeError: Value after * must be an iterable, not float